[web] Add DecodeWithEmbeddedJSONSchema function

ZFLloyd opened the pull request:
Add another decode function similar in functionality to `DecodeWithJSONschema` but taking a byte slice instead of a file path to make use of the file embedding.

julien-leclercq reviewed (commented):
Can you factorize code between the two functions please ?

ZFLloyd commented:
> Can you factorize code between the two functions please ?

I've refactored it as asked, but didn't completely extract the common part of the code, I've extracted the biggest part which was just validating the payload against the schema as it made the most sense to me.

Should we refactor it even more or is it fine this way ?

Also I didn't notice the tests for `web.DecodeWithJSONSchema()`at first, so I added the same tests for `web.DecodeWithEmbeddedJSONSchema` with some little adaptations to ensure it is working as intended with the embedding.

julien-leclercq commented:
> I've refactored it as asked, but didn't completely extract the common part of the code, I've extracted the biggest part which was just validating the payload against the schema as it made the most sense to me.
> 
> Should we refactor it even more or is it fine this way ?

I don't get it, beside the loading of the json schema the functions are doing the exact same job, just specialize the loading part no ? 

quickly did it so the `validateRequestPayload` might be wrong but something like 

```go
// DecodeWithJSONSchema takes the path to a json schema and a http request
// And returns an error when the request's payload does not match the JSON schema
func DecodeWithJSONSchema(request *http.Request, model interface{}, filePath string, options DecoderOptions) error {
	body, _ := ioutil.ReadAll(request.Body)

	_, rootFile, _, ok := runtime.Caller(1)
	if !ok {
		return fmt.Errorf("%w", NewErrInvalidJSONSchemaFilePath())
	}

	filePath = path.Join(path.Dir(rootFile), filePath)

	jsonSchema := gojsonschema.NewReferenceLoader("file://" + filePath)
	

	return validateRequestPayload(request, model, jsonSchema, payload)
}

// DecodeWithEmbeddedJSONSchema takes json schema and a http request
// And returns an error when the request's payload does not match the JSON schema
func DecodeWithEmbeddedJSONSchema(request *http.Request, model interface{}, jsonSchemaBytes []byte, options DecoderOptions) error {
	jsonSchema := gojsonschema.NewBytesLoader(jsonSchemaBytes)
	
	
	return validateRequestPayload(request, model, jsonSchema, payload);
}

func validateRequestPayload(request *http.Request, model interface{}, jsonSchema gojsonschema.JSONLoader, payload gojsonschema.JSONLoader) error {
	body, _ := ioutil.ReadAll(request.Body)
	payload := gojsonschema.NewBytesLoader(body)

	result, err := gojsonschema.Validate(jsonSchema, payload)
	if err != nil {
		return fmt.Errorf("%w: %v", NewErrBadRequestResponse(nil), err)
	}

	if !result.Valid() {
		requiredPropertyErrors := make(ErrorDetails)
		requestContentErrorDetails := make(ErrorDetails)

		for _, desc := range result.Errors() {
			propertyName := desc.Field()
			if propertyName == "(root)" {
				details, ok := desc.Details()["property"]
				if ok {
					propertyName = fmt.Sprintf("%v", details)
				}
			}

			if desc.Type() == "required" {
				requiredPropertyErrors[propertyName] = desc.Description()
				continue
			}

			requestContentErrorDetails[propertyName] = desc.Description()
		}

		if len(requiredPropertyErrors) > 0 {
			errDetails := requiredPropertyErrors
			for property, errMessage := range requestContentErrorDetails {
				errDetails[property] = errMessage
			}

			return fmt.Errorf("the request body is missing some required keys: %w", NewErrBadRequestResponse(errDetails))
		}

		return fmt.Errorf("%w", NewErrInvalidRequestBodyContent(requestContentErrorDetails))
	}
	
	request.Body = ioutil.NopCloser(bytes.NewBuffer(body))

	if err := decode(request, model, options); err != nil {
		return err
	}

	return nil
}
```

julien-leclercq reviewed (commented):
maybe we can mutualize the tests as well actually

ZFLloyd commented:
> maybe we can mutualize the tests as well actually



> > I've refactored it as asked, but didn't completely extract the common part of the code, I've extracted the biggest part which was just validating the payload against the schema as it made the most sense to me.
> > Should we refactor it even more or is it fine this way ?
> 
> I don't get it, beside the loading of the json schema the functions are doing the exact same job, just specialize the loading part no ?
> 
> quickly did it so the `validateRequestPayload` might be wrong but something like
> 
> ```go
> // DecodeWithJSONSchema takes the path to a json schema and a http request
> // And returns an error when the request's payload does not match the JSON schema
> func DecodeWithJSONSchema(request *http.Request, model interface{}, filePath string, options DecoderOptions) error {
> 	_, rootFile, _, ok := runtime.Caller(1)
> 	if !ok {
> 		return fmt.Errorf("%w", NewErrInvalidJSONSchemaFilePath())
> 	}
> 
> 	filePath = path.Join(path.Dir(rootFile), filePath)
> 
> 	jsonSchema := gojsonschema.NewReferenceLoader("file://" + filePath)
> 	
> 
> 	return validateRequestPayload(request, model, jsonSchema, payload)
> }
> 
> // DecodeWithEmbeddedJSONSchema takes json schema and a http request
> // And returns an error when the request's payload does not match the JSON schema
> func DecodeWithEmbeddedJSONSchema(request *http.Request, model interface{}, jsonSchemaBytes []byte, options DecoderOptions) error {
> 	jsonSchema := gojsonschema.NewBytesLoader(jsonSchemaBytes)
> 	
> 	
> 	return validateRequestPayload(request, model, jsonSchema, payload);
> }
> 
> func validateRequestPayload(request *http.Request, model interface{}, jsonSchema gojsonschema.JSONLoader, payload gojsonschema.JSONLoader) error {
> 	body, _ := ioutil.ReadAll(request.Body)
> 	payload := gojsonschema.NewBytesLoader(body)
> 
> 	result, err := gojsonschema.Validate(jsonSchema, payload)
> 	if err != nil {
> 		return fmt.Errorf("%w: %v", NewErrBadRequestResponse(nil), err)
> 	}
> 
> 	if !result.Valid() {
> 		requiredPropertyErrors := make(ErrorDetails)
> 		requestContentErrorDetails := make(ErrorDetails)
> 
> 		for _, desc := range result.Errors() {
> 			propertyName := desc.Field()
> 			if propertyName == "(root)" {
> 				details, ok := desc.Details()["property"]
> 				if ok {
> 					propertyName = fmt.Sprintf("%v", details)
> 				}
> 			}
> 
> 			if desc.Type() == "required" {
> 				requiredPropertyErrors[propertyName] = desc.Description()
> 				continue
> 			}
> 
> 			requestContentErrorDetails[propertyName] = desc.Description()
> 		}
> 
> 		if len(requiredPropertyErrors) > 0 {
> 			errDetails := requiredPropertyErrors
> 			for property, errMessage := range requestContentErrorDetails {
> 				errDetails[property] = errMessage
> 			}
> 
> 			return fmt.Errorf("the request body is missing some required keys: %w", NewErrBadRequestResponse(errDetails))
> 		}
> 
> 		return fmt.Errorf("%w", NewErrInvalidRequestBodyContent(requestContentErrorDetails))
> 	}
> 	
> 	request.Body = ioutil.NopCloser(bytes.NewBuffer(body))
> 
> 	if err := decode(request, model, options); err != nil {
> 		return err
> 	}
> 
> 	return nil
> }
> ```

I felt like it was two separate steps at first, but I've applied your suggestion and it doesn't seem as bad as it seemed to me at first 👍

ZFLloyd commented:
> maybe we can mutualize the tests as well actually

I've come up with something that doesn't look too bad. There's just one case that didn't fit both functions (the invalid file path) so I had to get another test into the Test function for `web.DecodeWithJSONSchema`.

The diff is really ugly though.

julien-leclercq commented on platform/web/tests/decode_with_json_schema_test.go:
```diff
@@ -10,190 +11,161 @@ import (
+	},
+}
+
+//go:embed testdata/json_schema_with_definition.json
+var jsonSchema []byte
```
I think this one can go in the function just below

ZFLloyd replied:
We can't use embed inside of functions or this happens:

```
# github.com/fewlinesco/go-pkg/platform/web/tests [github.com/fewlinesco/go-pkg/platform/web/tests.test]
./decode_with_json_schema_test.go:136:4: go:embed cannot apply to var inside func
```

I just found out trying to do so 😅

julien-leclercq replied:
Okay :) I guessed it was a possibility

julien-leclercq commented on platform/web/tests/decode_with_json_schema_test.go:
```diff
@@ -202,4 +174,58 @@ func TestDecodeWithJSONSchema(t *testing.T) {
 		})
 	}
+
+	t.Run("it returns an error when the file path is incorrect", func(t *testing.T) {
+		body := `{"code": "code", "id": "c9ecb26a-20ab-4acb-b34e-444457b06b3b", "datatype": "string"}`
+		request := prepareRequest(t, body)
+		var result expectedModel
+		if err := web.DecodeWithJSONSchema(request, &result, "testdata/json_schema/json_schema_with_definition.json", web.DecoderOptions{}); err != nil {
+			checkError(t, web.NewErrBadRequestResponse(nil), err)
+			return
+		}
+		t.Fatalf("the request body passed the validation but it should have returned an error")
+	})
```
can you extract it to another test function ? :)

julien-leclercq reviewed (commented):
I would have tested with the two decode functions in the same function to save the need for a private checkErrors, but this is also fine, not really blocking changes anymore
